### PR TITLE
lib/ffmpeg/encoderbase: add Intra refresh cases

### PR DESCRIPTION
**Pull request by fanli-hub**
Add Intra refresh cases to lib/ffmpeg/encoderbase.
Signed-off-by: Li Fan <fan1x.li@intel.com>

### CODE REVIEW

**Comment by fanli-hub** [comment]
> CI failed result for GST not this part change cause

Mengke is helping to add this part into gst. And I will add the parameter to gst cmd after Mengke completes it.

**Comment by uartie** [comment]
where are the tests that use this feature?

**Comment by fanli-hub** [comment]
"intra refresh" works when gop > int_ref_cycle_dist > int_ref_cycle_size. The value of cqp doesn't match the rule.

File: lib/ffmpeg/encoderbase.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ def gen_output_opts(self):
     if vars(self).get("vforced_idr", None) is not None:
       opts += " -force_key_frames expr:1 -forced_idr 1"
 
+    if self.codec.startswith("avc") and self.rcmode in ["cbr", "vbr", "cbr_lp", "vbr_lp"]:
+      opts += " -int_ref_type 1 -int_ref_cycle_size 4 -int_ref_cycle_dist 20"
+
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:92]: 1. There is no `cbr_lp` or `vbr_lp` for `self.rcmode`.  The lp is only captured in `self.lowpower`.  Thus, only need to check for `cbr` and `vbr` if want to apply to both vme and vdenc.
2. These properties are only valid for ffmpeg-qsv.  This adds them for all ffmpeg plugins, which is not correct.
3. Both AVC and HEVC support these properties in ffmpeg-qsv.  So why only AVC enabled?
4. This is essentially the same hack attempted earlier in the config file patch, just in another disguise.  As explained there, this adds these (hard-coded) parameters for all test cases in these categories.  Thus, this changes pre-existing cases and invalidates everyone's baseline refs for all test suites (quality, sanity, bat, full,...).  This is not acceptable.

Please create a new test for intra refresh and configure separate cases for them in the test configs.  Look at `test/ffmpeg-qsv/encode/avc.py::forced_idr` test for an example.  And the test function should include the intra refresh parameters that will be configured by user test config/spec.

**Review by uartie — CHANGES_REQUESTED** [review]
unacceptable

File: lib/ffmpeg/encoderbase.py [hunk 2]
```diff
@@ -86,6 +86,8 @@ def gen_output_opts(self):
       opts += " -look_ahead_depth {ladepth}"
     if vars(self).get("vforced_idr", None) is not None:
       opts += " -force_key_frames expr:1 -forced_idr 1"
+    if vars(self).get("vintr_ref", None) is not None:
+      opts += " -int_ref_type 1 -int_ref_cycle_size 4 -int_ref_cycle_dist 20"
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:90]: Why hardcoded values?  Maybe better to let test config set these parameter values.

Comment (uartie) [inline lib/ffmpeg/encoderbase.py:90]: Why hardcoded values?  Maybe better to let test config set these parameter values.

Comment (fanli-hub) [reply lib/ffmpeg/encoderbase.py]: > Why hardcoded values? Maybe better to let test config set these parameter values.

How about doing like this time?
Moreover, hevc "-low_power 0" doesn't support Intra refresh, so I only add intra refresh when hevc "-low_power 1".

File: lib/ffmpeg/encoderbase.py [hunk 3]
```diff
@@ -86,6 +86,12 @@ def gen_output_opts(self):
       opts += " -look_ahead_depth {ladepth}"
     if vars(self).get("vforced_idr", None) is not None:
       opts += " -force_key_frames expr:1 -forced_idr 1"
+    if vars(self).get("ref_type", None) is not None:
+      opts += " -int_ref_type {ref_type}"
+    if vars(self).get("ref_type", None) is not None:
+      opts += " -int_ref_cycle_size {ref_size}"
+    if vars(self).get("ref_type", None) is not None:
+      opts += " -int_ref_cycle_dist {ref_dist}"
```
Comment (uartie) [inline lib/ffmpeg/encoderbase.py:94]: copy/paste error... all these conditionals are exactly the same.

The intra refresh type, size, and dist depend on each other.  That is, it does not make sense to have one but not the others.  Therefore, how about a variable like `intref = dict(type = T, size = S, dist = D)`.  Then do:

```python
if vars(self).get("intref", None) is not None:
  opts += " -int_ref_type {intref[type]}"
  opts += " -int_ref_cycle_size {intref[size]}"
  opts += " -int_ref_cycle_dist {intref[dist]}"
```

File: test/ffmpeg-qsv/encode/avc.py [hunk 4]
```diff
@@ -240,3 +240,49 @@ def init(self, tspec, case, rcmode, bitrate, maxrate, qp, quality, profile):
   def test(self, case, rcmode, bitrate, maxrate, qp, quality, profile):
     self.init(spec, case, rcmode, bitrate, maxrate, qp, quality, profile)
     self.encode()
+
+class intr_ref(AVCEncoderTest):
```
Comment (uartie) [inline test/ffmpeg-qsv/encode/avc.py:244]: How about `class intref` instead?

File: test/ffmpeg-qsv/encode/avc.py [hunk 5]
```diff
@@ -240,3 +240,49 @@ def init(self, tspec, case, rcmode, bitrate, maxrate, qp, quality, profile):
...
+  def init(self, tspec, case, gop, bframes, bitrate, qp, maxrate, profile, rcmode, ref_type, ref_size, ref_dist):
```
Comment (uartie) [inline test/ffmpeg-qsv/encode/avc.py:245]: How about `reftype, refsize, refdist` param names instead?

File: test/ffmpeg-qsv/encode/avc.py [hunk 6]
```diff
@@ -240,3 +240,49 @@ def init(self, tspec, case, rcmode, bitrate, maxrate, qp, quality, profile):
...
+      ref_size   = ref_size,
+      ref_dist   = ref_dist,
+    )
+
+  @slash.parametrize(*gen_avc_intr_ref_parameters(spec, ['high', 'main', 'baseline']))
+  def test(self, case, gop, bframes, bitrate, qp, maxrate, profile, rcmode, ref_type, ref_size, ref_dist):
+    self.init(spec, case, gop, bframes, bitrate, qp, maxrate, profile, rcmode, ref_type, ref_size, ref_dist)
+    self.encode()
+
+class intr_ref_lp(AVCEncoderLPTest):
```
Comment (uartie) [inline test/ffmpeg-qsv/encode/avc.py:267]: `class intref_lp`

File: test/ffmpeg-qsv/encode/avc.py [hunk 7]
```diff
@@ -240,3 +240,49 @@ def init(self, tspec, case, rcmode, bitrate, maxrate, qp, quality, profile):
...
+      bitrate    = bitrate,
+      case       = case,
+      maxrate    = maxrate,
+      minrate    = bitrate,
+      profile    = profile,
+      qp         = qp,
+      gop        = gop,
+      ref_type   = ref_type,
+      ref_size   = ref_size,
+      ref_dist   = ref_dist,
```
Comment (uartie) [inline test/ffmpeg-qsv/encode/avc.py:259]: With other suggestions, do:
```python
   intref = dict(type = reftype, size = refsize, dist = refdist)
```

**Review by uartie — CHANGES_REQUESTED** [review]
almost there ;)

File: lib/ffmpeg/encoderbase.py (hunk 2 above)
Comment (uartie) [reply lib/ffmpeg/encoderbase.py]: ok, good.

File: test/ffmpeg-qsv/encode/avc.py (hunk 7 above)
Comment (fanli-hub) [reply test/ffmpeg-qsv/encode/avc.py]: Thank you for your patience and detailed reply. (●'◡'●)

**Pull request merged**